src/python/WMCore/Services/DBS/DBS3Reader.py

```diff
@@ -1016,44 +1043,39 @@ def fixMissingParentageDatasets(self, childDataset, insertFlag=True):
             try:
-                numFiles = self.findAndInsertMissingParentage(blockName, insertFlag=insertFlag)
+                self.logger.info("Fixing parentage for block: %s", blockName)
+                numFiles = self.findAndInsertMissingParentage(blockName, parentFullInfo, insertFlag=insertFlag)
                 self.logger.debug("%s file parentage added for block %s" % (numFiles, blockName))
             except Exception as ex:
                 self.logger.exception("Parentage updated failed for block %s", blockName)
                 failedBlocks.append(blockName)
 
         return failedBlocks
 
-    def insertMissingParentageForAllFiles(self, childDataset, filterFilesWithParents=True, insertFlag=False):
-        """
-        :param childDataset: child dataset need to set the parentage correctly.
-        :param filterFilesWithParents: if True, only select files without parents, if False all the files in the dataset
-        :param insertFlag: if True, insert to DBS, if False just get the list of the file parentage without insert
-        :return: blocks which failed to insert parentage. should be used for retrying
-        """
-        blocks = [b['block_name'] for b in self.dbs.listBlocks(dataset=childDataset)]
-        failedBlocks = []
-        print("Handling %d blocks" % len(blocks))
-        totalFiles = 0
-        for blockName in blocks:
-            try:
-                if filterFilesWithParents:
-                    childLFNs = self.listFilesWithNoParents(blockName)
-                    if len(childLFNs) == 0:
-                        continue
-                else:
-                    childLFNs = []
-
-                numFiles = self.findAndInsertMissingParentage(blockName, childLFNs=childLFNs, insertFlag=insertFlag)
-                print("%s file parentage added for block %s" % (numFiles, blockName))
-                totalFiles += numFiles
-            except Exception as e:
-                print(traceback.format_exc())
-                failedBlocks.append(blockName)
-        print("Total pairs: ", totalFiles)
-        return failedBlocks
+    def getParentDatasetTrio(self, childDataset):
+        """
+        Provided a dataset name, return all the parent dataset information, such as:
+          - file ids, run number and lumi section
+        NOTE: This API is meant to be used by the StepChainParentage thread only!!!
+        :param childDataset: name of the child dataset
+        :return: a dictionary where the key is a set of run/lumi, its value is the fileid
+        """
+        # this will return data in the format of:
+        # {'554307997': [[1, 557179], [1, 557178],...
+        # such that: key is file id, in each list is [run_number, lumi_section_numer].
+        parentFullInfo = self.dbs.listParentDSTrio(dataset=childDataset)
+
+        # runs the actual mapping logic, like {"child_id": ["parent_id", "parent_id2", ...], etc
+        parentFrozenData = {}
+        for item in parentFullInfo:
```

Comment:
Same as above.

Reply:
Output of `listParentDSTrio` API is a list of dictionaries (AFAIK only one dictionary though). So I'm afraid there isn't much we can do.